morango/utils/syncsession.py

```diff
@@ -0,0 +1,111 @@
+import uuid
+
+from django.db import connection, transaction
+from django.forms.models import model_to_dict
+from django.utils.six import iteritems
+from morango.models import DataTransferBuffer, RecordMaxCounter, RecordMaxCounterBuffer, Store, SyncSession
+
+
+def _join_with_logical_operator(lst, operator):
+    op = ") {operator} (".format(operator=operator)
+    return "(({items}))".format(items=op.join(lst))
+
+
+class SyncConnection(object):
+
+    def get_remote_certifcates():
+        pass
+
+    def get_local_certificates():
+        pass
+
+    def start_sync(local_certificate, remote_certificate=None):
+        pass
+
+    def request_certificate(username, password):
+        pass
+
+    def provide_certificate(role):
+        pass
+
+
+class SyncController(object):
+
+    def __init__(self, host, profile):
+        self.host = host
+        self.profile = profile
+        self.session_id = uuid.uuid4()
+        self.transfer_session_id = None
+        self.symmetric_key = 'xxx'
+        self.sync_session = SyncSession(host=host, id=self.session_id, symmetric_key=self.symmetric_key)
+
+    def initiate_push_request(self, sync_filter, chunksize):
+        pass
+
+    def initiate_pull_request(self, sync_filter, fsics, transfer_id):
+        pass
+
+    def close(self):
+        pass
+
+    @transaction.atomic
+    def _queue_into_buffer(self, fsics):
+        """
+        Takes a chunk of data from the store to be put into the buffer to be sent to another morango instance.
+        """
+        where_statements = []
+        # create condition for all incoming FSICs where instance_ids are equal, but internal counters are higher then FSICs counters
+        for instance, counter in iteritems(fsics):
+            where_statements += ['(last_saved_instance == {0} AND last_saved_counter > {1})'.format(instance, counter)]
```

Comment:
We also need to make sure (doesn't need to be here, necessarily -- can possibly be when we calculate the FSICs) that we include all instances that the other side doesn't even know about -- i.e. the "implicit counter=0" instances.